File: training/swift/Sources/CellarsAndCentaurs/Domain/Model/HitPoints.swift
/// Hit points drop when the creature takes damage and recover when it rests in its lair.
struct HitPoints {
    static let MAX_HIT_POINTS = 999

    @NonNegative var current: Int
    let maximum: Int
    var temporary: Int

    init(current: Int, maximum: Int, temporary: Int = 0) {
        self.current = current
        self.maximum = min(maximum, HitPoints.MAX_HIT_POINTS)
        self.temporary = temporary
    }

    static func full(_ maximum: Int) -> HitPoints {
        HitPoints(current: maximum, maximum: maximum)
    }

    func damaged(by amount: Int) -> HitPoints {
        HitPoints(current: current - amount, maximum: maximum, temporary: temporary)
    }

    func rested() -> HitPoints {
        HitPoints.full(maximum)
    }
}

File: training/csharp/src/De.Sots.CellarsAndCentaurs/Domain/Model/HitPoints.cs
namespace De.Sots.CellarsAndCentaurs.Domain.Model;

/// <summary>
/// Hit points drop when the creature takes damage and recover when it rests in its lair.
/// </summary>
public sealed class HitPoints
{
    public const int MAX_HIT_POINTS = 999;

    public int Current { get; private set; }

    public int Max { get; }

    public int Temporary { get; private set; }

    public HitPoints(int current, int max, int temporary = 0)
    {
        Current = current;
        Max = Math.Min(max, MAX_HIT_POINTS);
        Temporary = temporary;
    }

    public static HitPoints Init(int max) => new(max, max);

    public HitPoints Damaged(int damage) => new(Math.Max(0, Current - damage), Max, Temporary);

    public HitPoints Rested() => new(Max, Max, Temporary);
}
